### PR TITLE
#279-ci-cd-setup

### PR DESCRIPTION
**Pull request by lewisdaly**
- Set up E2E tests with jest, steps taken from postman collection
- tidy up the newman scripts runner, and make newman possible to run without a global install
- set up npm project

### CODE REVIEW

File: scripts/_wait4_all.js [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+ * @function isSystemFailing
+ * @param {*} waitingMap
+ * @returns {boolean}
+ */
+function isSystemFailing (waitingMap) {
+  const unhealthyCount = Object
+    .keys(waitingMap)
+    .filter(k => waitingMap[k] === 'unhealthy').length
+
+  if (unhealthyCount > 0) {
```
Comment (eoln) [inline scripts/_wait4_all.js:125]: what about `return unhealthyCount > 0`

File: scripts/_wait4_all.js [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+  return result
+}
+
+/**
+ * @function isSystemHealthy
+ * @param {*} waitingMap
+ * @returns {boolean}
+ */
+function isSystemHealthy (waitingMap) {
+  const healthyCount = Object
```
Comment (eoln) [inline scripts/_wait4_all.js:108]: what about extracting filtering logic into reusable function?

File: scripts/_wait4_all.js [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+/**
+ * @function getProgress
+ * @description Invokes the `docker inspect` command for the given container
+ * @param {string} containerName
+ * @returns {'healthy' | 'unhealthy' | 'starting'}
+ */
+function getProgress (containerName) {
+  const command = `docker inspect --format='{{json .State.Health.Status}}' ${containerName}`
+  const result = execSync(command).toString().replace(/['"]+|[\n]+/g, '')
```
Comment (eoln) [inline scripts/_wait4_all.js:97]: simply return?

File: scripts/_wait4_all.js [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+async function main () {
+  const waitingMap = {}
+  // serviceName => status, where status is unknown, healthy, unhealthy or starting
+  expectedContainers.forEach(serviceName => {
+    waitingMap[serviceName] = 'unknown'
+  })
+
+  try {
+    while (retries > 0) {
+      await updateServiceStatus(waitingMap)
```
Comment (eoln) [inline scripts/_wait4_all.js:36]: what about moving this block into a separate function, so the while-loop will be easy to understand?

**Comment by lewisdaly** [comment]
> It is Cool! @lewisdaly 👍 Nice cleanup and very fine setup for ts driven tests. Few not very important comments. The data preparation sections are elaborative ... ;-) What about extracting data to external files... ?

Thanks @eoln . I'll touch up the `_wait4_all.js` script as you suggest. 

For the "data preparation sections", do you mean within the tests themselves? If so, I do agree that the data preparation stuff can live elsewhere. I mainly just didn't want to prematurely optimize especially since we don't know exactly what form these tests will take at the moment.

Indeed I can imagine some sort of TestClient library we will write in the future to invoke the http requests with some nice helpers etc.

**Pull request merged**